extensions/cl_ext_buffer_device_address.asciidoc

```diff
@@ -0,0 +1,289 @@
+application. This is not an issue in systems which can provide virtual memory
+across the platform, but might provide implementation challenges in cases
+where the device presents a global memory with its disjoint address space
+(that can also be a physical memory address space) or, for example, when
+a barebone embedded system lacks virtual memory support altogether.
+
+Various higher-level APIs present a memory allocation routine which can
+allocate device-only memory and provide raw pointers to it without guarentees
+of system-wide uniqueness: Minimal implementations of OpenMP's omp_target_alloc() and
+CUDA/HIP's cudaMalloc()/hipMalloc() do not require a shared
+address space between the host and the device. This extension is meant to
+provide a minimal set of features to implement such APIs without requiring
+a shared virtual address space between the host and the device.
+
+=== New API Function
+
+include::{generated}/api/protos/clSetKernelArgDevicePointerEXT.txt[]
+
+=== New API Enums
+
+Enums for enabling device pointer properties when creating a buffer
+{clCreateBuffer}, see <<clCreateBuffer, the list of supported memory flag values table>>:
+
+[source]
+----
+#define CL_MEM_DEVICE_ADDRESS_EXT (1ul << 31)
+#define CL_MEM_DEVICE_PRIVATE_EXT (1ul << 30)
+----
+
+Enums for querying the device pointer from the cl_mem <<clGetMemObjectInfo, the list of supported param_names table>>:
+
+[source]
+----
+#define CL_MEM_DEVICE_PTR_EXT 0xff01
+----
+
+Enums for setting information of indirect device pointer accesses to kernels <<clSetKernelExecInfo, the list of supported param_names table>>. This is for OpenCL 2.0 and above. When implementing the
+extension on an older OpenCL version, indirect device pointer access is not supported.
+
+[source]
+----
+#define CL_KERNEL_EXEC_INFO_DEVICE_PTRS_EXT 0x11B8
+----
+
+== New API Types
+
+Returned as the query result value *clGetMemObjectInfo* with `CL_DEVICE_PTR_EXT`.
+
+[source]
+----
+typedef cl_ulong cl_mem_device_address_EXT;
+----
+
+Returned as the query result value *clGetMemObjectInfo* with `CL_DEVICE_PTRS_EXT`.
+
+[source]
+----
+typedef struct _cl_mem_device_address_pair_EXT
+{
+  cl_device_id device;
+  cl_mem_device_address_EXT address;
+} cl_mem_device_address_pair_EXT;
+----
+
+== Modifications to the OpenCL API Specification
+
+=== Section 5.2.1 - Creating Buffer Objects:
+
+Add new allocation flags <<clCreateBuffer, List of supported memory flag values table>>:
+
+[[list-of-supported-memory-flag-values-adds]]
+.List of supported memory flags by {clCreateBuffer}
+[width="100%",cols="<50%,<50%",options="header"]
+|====
+| Memory Flags | Description
+| {CL_MEM_DEVICE_ADDRESS_EXT_anchor}
+
+include::{generated}/api/version-notes/CL_MEM_DEVICE_ADDRESS_EXT.asciidoc[]
+  | This flag specifies that the buffer must have a single fixed address
+   for its lifetime and the address should be unique at least across the devices
+   of the context, but not necessarily withing the host (virtual) memory.
```

Comment:
I think support for this flag needs to be optional, as otherwise this extension will be difficult to implement on current Vulkan implementations.

With Vulkan you can use bda (optional in 1.2, mandatory in 1.3) to have a fixed address for a memory allocation, but a CL runtime wouldn't be able to guarantee multiple devices to get the same address.

Reply:
or maybe it's better if it's optional for multi device contexts, so applications and libraries could support the same code on single device contexts as well.